src/storage-client/src/controller.rs

```diff
@@ -1383,6 +1538,17 @@ where
     }
 
     fn drop_sources_unvalidated(&mut self, identifiers: Vec<GlobalId>) {
+        for id in identifiers.iter() {
```

Comment:
I think this should also remove the dependencies from the collection when it drops the read capabilities, e.g. something like:

```diff
diff --git a/src/storage-client/src/controller.rs b/src/storage-client/src/controller.rs
index 2076d7971..d1681746c 100644
--- a/src/storage-client/src/controller.rs
+++ b/src/storage-client/src/controller.rs
@@ -1614,13 +1614,13 @@ where
 
     fn drop_sources_unvalidated(&mut self, identifiers: Vec<GlobalId>) {
         for id in identifiers.iter() {
-            let collection = match self.collection(id.clone()) {
+            let collection = match self.collection_mut(*id) {
                 Ok(collection) => collection,
                 Err(_) => continue,
             };
 
             let read_capability = collection.read_capabilities.frontier().to_owned();
-            let storage_dependencies = collection.storage_dependencies.clone();
+            let storage_dependencies = collection.storage_dependencies.drain(..).collect_vec();
             self.remove_read_capabilities(read_capability, &storage_dependencies);
         }
```

because the calls to `remove_read_capabilities` and `set_read_policy` both call `update_read_capabilities`. 
- `remove_read_capabilities`removes the read capability from the dependency
-  `update_read_capabilities` tries to propagate the change to this ID's since to the dependency, even though we've just dropped the capability it holds.

This will result in a double-retraction and panic in in the new `assert` you added [here](https://github.com/MaterializeInc/materialize/pull/17603/files#diff-3154d32c4d0818b6513328628453395000c38bd1238bd1fa9e96b796201b38a6R1751-R1757).